packages/typescript-estree/src/parser.ts

```diff
@@ -235,12 +270,21 @@ function parseAndGenerateServices<T extends TSESTreeOptions = TSESTreeOptions>(
    */
   return {
     ast: estree as AST<T>,
-    services: {
-      hasFullTypeInformation: shouldProvideParserServices,
-      program,
-      esTreeNodeToTSNodeMap: astMaps.esTreeNodeToTSNodeMap,
-      tsNodeToESTreeNodeMap: astMaps.tsNodeToESTreeNodeMap,
-    },
+    services:
+      shouldProvideParserServices && program != null
+        ? {
+            program,
+            esTreeNodeToTSNodeMap: astMaps.esTreeNodeToTSNodeMap,
+            tsNodeToESTreeNodeMap: astMaps.tsNodeToESTreeNodeMap,
+          }
+        : {
+            program: null,
+            // we still return the node maps because
+            // (a) they don't require type info and
+            // (b) they can be useful when using some of TS's internal non-type-aware AST utils
+            esTreeNodeToTSNodeMap: astMaps.esTreeNodeToTSNodeMap,
+            tsNodeToESTreeNodeMap: astMaps.tsNodeToESTreeNodeMap,
+          },
```

Comment:
Nit: simplifying a bit...

```suggestion
    services: {
      program,
      // we always return the node maps even if there's no program because
      // (a) they don't require type info and
      // (b) they can be useful when using some of TS's internal non-type-aware AST utils
      esTreeNodeToTSNodeMap: astMaps.esTreeNodeToTSNodeMap,
      tsNodeToESTreeNodeMap: astMaps.tsNodeToESTreeNodeMap,
    },
```